### PR TITLE
Time stretching 5 of 6 repair effects and generators for stretched audio

### LINKED ISSUE #4850: Audio clip stretching
As opposed to now, changing project tempo must modify the clip lengths in the time view but not the beats-and-measure view.
Also, the stretch handles (todo refer to issue) should affect stretch ratio and hence clip length (in both views).
The painting of the waveform and spectrum must look convincing.

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #4850

This PR repairs destructive effects for stretched clips.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

## QA
All effects are supposed to work. Trying all of them may be too much, but a variety of types of effects should be tried (time-related effect, Nyquist, EQs and filters, noise removal, pitch shift, spectral tools, etc. Of course try them on stretched data, but also on stereo data, etc.

Note that applying an effect on a partial clip selection will create a new clip if the clip was stretched. That's wanted. The old behavior is only preserved on unstretched data.

- [ ] AutoDuck
- [ ] Change Speed
- [ ] Click Removal
- [ ] EQ
- [ ] Generate silence
- [ ] Loudness
- [ ] Noise Reduction
- [ ] Normalize
- [ ] Paulstretch
- [ ] Repair (Since this requires up to 128 samples only, the result will probably only be silence, and that's ok: it's a known limitation of the algorithm that rendering very tiny portions of audio only output silence)
- [ ] Repeat
- [ ] Reverse
- [ ] Change Tempo and Sliding Scale, high and low quality
- [ ] Stereo To Mono (previously reported by @dozzer as https://github.com/audacity/audacity/issues/5118)
- [ ] Truncate Silence
- [ ] Compressor
- [ ] Some Nyquist effects and generators
- [ ] At least one "Per Track Effect" example (any of the realtime capable, built-in or plug-in effects, but used destructively)

### CODE REVIEW

File: src/effects/nyquist/Nyquist.cpp [hunk 1]
```diff
@@ -729,7 +729,7 @@ bool NyquistEffect::Process(EffectInstance &, EffectSettings &settings)
    // to operate on the selected wave tracks
    std::optional<EffectOutputTracks> oOutputs;
    if (!bOnePassTool)
-      oOutputs.emplace(*mTracks, true);
+      oOutputs.emplace(*mTracks, std::pair<double, double> { mT0, mT1 }, true);
```
Comment (Paul-Licameli) [inline src/effects/nyquist/Nyquist.cpp:732]: You might simplify the expression

File: libraries/lib-track/Track.cpp [hunk 2]
```diff
@@ -988,7 +989,7 @@ TrackList::RegisterPendingChangedTrack(Updater updater, Track *src)
    TrackListHolder tracks;
    std::vector<Track*> result;
    if (src) {
-      tracks = src->Clone(); // not duplicate
+      tracks = src->Clone(std::nullopt); // not duplicate
```
Comment (Paul-Licameli) [inline libraries/lib-track/Track.cpp:992]: I think you can just write `{}` here and elsewhere that you wrote `nullopt`

**Comment by Paul-Licameli** [comment]
I'm reviewing the PRs out of order.  Some of them are smaller, more fresh in my mind from recent review of the closed branch, and less problematic.

This one looks good.  I added to the QA checklist.

**Review by Paul-Licameli — APPROVED** [review]

File: src/effects/nyquist/Nyquist.cpp (hunk 1 above)
Comment (Paul-Licameli) [reply src/effects/nyquist/Nyquist.cpp:772]: I repeat this

File: src/effects/Generator.cpp [hunk 3]
```diff
@@ -71,7 +71,7 @@ bool Generator::Process(EffectInstance &, EffectSettings &settings)
                   ViewInfo::Get(*pProject).selectedRegion;
                track.ClearAndPaste(
                   selectedRegion.t0(), selectedRegion.t1(),
-                  *list, true, false, &warper);
+                  *list, true, true, &warper);
```
Comment (Paul-Licameli) [inline src/effects/Generator.cpp:74]: Can you explain why this change is correct?

Comment (saintmatthieu) [reply src/effects/Generator.cpp:74]: `merge` will become `mergeIfPossible` in `PasteOne`. This preserves the old behavior if possible, i.e., if `selectedRegion` spans clips of stretch ratio 1.

File: src/effects/nyquist/Nyquist.cpp (hunk 1 above)
Comment (saintmatthieu) [reply src/effects/nyquist/Nyquist.cpp:772]: I haven't found how ...

File: src/effects/Generator.cpp (hunk 3 above)
Comment (Paul-Licameli) [reply src/effects/Generator.cpp:74]: I don't accept that explanation.

Compare behavior before and after this PR:
- Generate noise.
- Split the clip.
- Select around the split.
- Generate silence.

There is a change of behavior, as shown below.  But I thought the intention, with effects and generators (unlike with copy/paste), and no stretches, was to change no behavior.

Result, before this PR:

image

After:

image

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

**Comment by Paul-Licameli** [comment]
I think the correct solution may require a third boolean argument.  The `merge` argument in master was never passed into `PasteOne` and governed something other than what happens and the edges of the insertion.

**Comment by saintmatthieu** [comment]
Things to fix on this branch:
* Auto Duck
* Generate silence over stretched clip
* Noise reduction
* Change tempo
* Change pitch
* Stereo to mono
* Nyquist noise gate (and other nyquist effects probably too)

File: src/effects/nyquist/Nyquist.cpp (hunk 1 above)
Comment (Paul-Licameli) [reply src/effects/nyquist/Nyquist.cpp:772]: I was wrong!  It's `emplace` and not a direct use of the contstructor

**Comment by Paul-Licameli** [comment]
I think we can agree to close the PR, as #5138 is its replacement.
